File: typescript/src/lib/execution/executable_module_service_impl.ts
import { ServerUnaryCall, sendUnaryData } from "grpc";
import { EnclaveContext, ExecuteArgs, ExecuteResponse, IExecutableModuleServiceServer } from "kurtosis-core-api-lib";
import { KnownKeysOnly } from "minimal-grpc-server";
import { newExecuteResponse } from "../constructor_calls";
import { ExecutableKurtosisModule } from "../kurtosis_modules/executable_kurtosis_modules";
import * as google_protobuf_empty_pb from "google-protobuf/google/protobuf/empty_pb";

// The KnownKeysOnly thing is a workaround due to a silly gRPC requirement that your service implement
//  UnimplementedServiceServer
export class ExecutableModuleServiceImpl implements KnownKeysOnly<IExecutableModuleServiceServer> {
    constructor(
        private readonly module: ExecutableKurtosisModule,
        private readonly enclaveCtx: EnclaveContext,
    ) {}

    public isAvailable(call: ServerUnaryCall<google_protobuf_empty_pb.Empty>, callback: sendUnaryData<google_protobuf_empty_pb.Empty>): void {
        callback(null, new google_protobuf_empty_pb.Empty());
    }

    public execute(call: ServerUnaryCall<ExecuteArgs>, callback: sendUnaryData<ExecuteResponse>): void {
        const args: ExecuteArgs = call.request;
        this.module.execute(
            this.enclaveCtx, 
            args.getParamsJson()
        ).then(executeResult => {
            if (executeResult.isErr()) {
                callback(executeResult.error, null);
                return;
            }

            const responseJson: string = executeResult.value;
            const executeResponse: ExecuteResponse = newExecuteResponse(responseJson);

            callback(null, executeResponse);
        })
    }
}

File: golang/lib/execution/executable_module_service_impl.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package execution

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis-module-api-lib/golang/kurtosis_module_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-module-api-lib/golang/lib/kurtosis_modules"
	"github.com/kurtosis-tech/stacktrace"
	"google.golang.org/protobuf/types/known/emptypb"
)

type executableModuleServiceImpl struct {
	// This embedding is required by gRPC
	kurtosis_module_rpc_api_bindings.UnimplementedExecutableModuleServiceServer
	module     kurtosis_modules.ExecutableKurtosisModule
	enclaveCtx *enclaves.EnclaveContext
}

func newExecutableModuleServiceImpl(
	module kurtosis_modules.ExecutableKurtosisModule,
	enclaveCtx *enclaves.EnclaveContext,
) *executableModuleServiceImpl {
	return &executableModuleServiceImpl{
		module:     module,
		enclaveCtx: enclaveCtx,
	}
}

func (server *executableModuleServiceImpl) IsAvailable(ctx context.Context, empty *emptypb.Empty) (*emptypb.Empty, error) {
	return &emptypb.Empty{}, nil
}

func (server *executableModuleServiceImpl) Execute(ctx context.Context, args *kurtosis_module_rpc_api_bindings.ExecuteArgs) (*kurtosis_module_rpc_api_bindings.ExecuteResponse, error) {
	result, err := server.module.Execute(server.enclaveCtx, args.ParamsJson)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred executing the module")
	}
	executeResponse := &kurtosis_module_rpc_api_bindings.ExecuteResponse{
		ResponseJson: result,
	}

	return executeResponse, nil
}
